### PR TITLE
Release 1.2.0

### PR DESCRIPTION
**Pull request by hellais**
I am going to use this branch to work towards releasing 1.2.0.

Checklist for release:

* [x] Read diff with some extra care
* [x] Smoke test it with some of my software that uses pypcap
* [x] Do a bit of manual testing with python2 and python3
* [x] Update changelog entry mentioning all the bugs that have been closed and added python 3 support
* [x] Bump version number
* [ ] Tag a release
* [ ] Upload to pypi

### CODE REVIEW

File: pcap.pyx [hunk 1]
```diff
@@ -144,81 +147,88 @@ dltoff = { DLT_NULL:4, DLT_EN10MB:14, DLT_IEEE802:22, DLT_ARCNET:6,
...
     cdef char *__filter
     cdef char __ebuf[256]
     cdef int __dloff
-    
+
     def __init__(self, name=None, snaplen=65535, promisc=True,
                  timeout_ms=0, immediate=False):
         global dltoff
         cdef char *p
-
```
Comment (hellais) [inline pcap.pyx]: note to self. Ensure spacing is consistent across the all files.

File: pcap.pyx [hunk 2]
```diff
@@ -227,92 +237,85 @@ cdef class pcap:
...
+
         cnt      -- number of packets to process;
                     or 0 to process all packets until an error occurs,
                     EOF is reached, or the read times out;
                     or -1 to process all packets received in one buffer
         callback -- function with (timestamp, pkt, *args) prototype
         *args    -- optional arguments passed to callback on execution
         """
-        cdef pcap_handler_ctx ctx
+        cdef pcap_handler_ctx ctx = pcap_handler_ctx()
```
Comment (hellais) [inline pcap.pyx]: 👍

File: pcap.pyx [hunk 3]
```diff
@@ -322,20 +325,21 @@ cdef class pcap:
...
-                callback(hdr.ts.tv_sec + (hdr.ts.tv_usec / 1000000.0),
-                         PyBuffer_FromMemory(pkt, hdr.caplen), *args)
+                callback(
+                    hdr.ts.tv_sec + (hdr.ts.tv_usec / 1000000.0),
+                    get_buffer(pkt, hdr.caplen),
+                    *args
+                )
             elif n == 0:
-                break
+                continue
```
Comment (hellais) [inline pcap.pyx]: TODO: remember to update changelog to mention that this fixes bug in handling all packets when `cnt=0` is set.

File: pcap.pyx [hunk 4]
```diff
@@ -227,92 +237,85 @@ cdef class pcap:
...
 
     def loop(self, cnt, callback, *args):
         """Processing packets with a user callback during a loop.
         The loop can be exited when cnt value is reached
         or with an exception, including KeyboardInterrupt.
-        
+
         Arguments:
 
         cnt      -- number of packets to process;
```
Comment (hellais) [inline pcap.pyx]: TODO: update docstring to mention that actually only `cnt=0` (and not also `-1`) will process all packets until an error occurs or EOF is reached.

File: pcap.pyx [hunk 5]
```diff
@@ -369,22 +373,23 @@ cdef class pcap:
...
+                    hdr.ts.tv_sec + (hdr.ts.tv_usec / 1000000.0),
+                    get_buffer(pkt, hdr.caplen),
+                )
             elif n == 0:
-                return None
+                continue
             elif n == -1:
                 raise KeyboardInterrupt
             elif n == -2:
                 raise StopIteration
```
Comment (hellais) [inline pcap.pyx]: There is a fair amount of duplication between what is in here and what is in loop. Maybe they can be refactored to be consolidated.

File: pcap.pyx (hunk 3 above)
Comment (hellais) [inline pcap.pyx]: I see this changes the behavior with how timeouts are handled. TODO: mention this in the changelog.

**Comment by hellais** [comment]
@nuald do you recall how you ran cython on the `.pyx` file. I don't seem to be able to get the exact same `.c` file as output. Any ideas how we can make this "reproducible"?

**Comment by nuald** [comment]
I guess you see this Cython issue: [encodes absolute source path in output](https://github.com/cython/cython/issues/1565). Cython hadn't have releases for awhile and the fix is not in the stable version, so I'd recommend to install it from the sources (as it was mentioned in the documentation change I've submitted within Python 3 support code).

**Comment by hellais** [comment]
> I guess you see this Cython issue: encodes absolute source path in output. Cython hadn't have releases for awhile and the fix is not in the stable version, so I'd recommend to install it from the sources (as it was mentioned in the documentation change I've submitted within Python 3 support code).

Yes that is one issue, but still even with the latest cython (0.26b0) I get a different `.c` file in output.

**Comment by nuald** [comment]
Yes, the C output differs, but I've verified - the newly generated C code is fine, all tests for both Py2 and Py3 are passed. You may want to regenerate C code and use it, or merge the changes from my branch (https://github.com/nuald/pypcap/tree/python3) - I've submitted the regenerated C code there.

**Comment by wasilukm** [comment]
Hi, what is a timeline of releasing 1.2.0?

**Comment by hellais** [comment]
Hi @wasilukm sorry I ended up being quite busy the past months. I will see if I can get something out this week.

**Comment by hellais** [comment]
So I made a release candiate that includes this branch: https://github.com/pynetwork/pypcap/releases/tag/v1.2.0-rc.1

 I am going to give people a bit of time to test it and possibly report bugs and if nothing happens I will cut 1.2.0 and upload it to pypi.

Sound reasonable?

**Comment by nuald** [comment]
I guess so. However, installation instructions are out-dated (Npcap team pulled their project out from the Chocolatey). Do you want me to update it, and if yes, in what branch?

**Comment by hellais** [comment]
> However, installation instructions are out-dated (Npcap team pulled their project out from the Chocolatey). Do you want me to update it, and if yes, in what branch?

yes please, that would be very useful.

I am not a Windows user so I am probably not the best to do that.

You should create a new PR with the documentation updates.

**Pull request merged**